lib/elixir/lib/calendar/time_zone_database.ex

```diff
@@ -0,0 +1,89 @@
+defmodule Calendar.TimeZoneDatabase do
+  @moduledoc """
+  This module defines a behaviour for providing time zone data.
+
+  IANA provides time zone data that includes data about different
+  UTC offsets and standard offsets for timezones.
+  """
+
+  @typedoc """
+  A period where a certain combination of UTC offset, standard offset and zone
+  abbreviation is in effect.
+
+  For instance one period could be the summer of 2018 in "Europe/London" where summer time /
+  daylight saving time is in effect and lasts from spring to autumn. At autumn the `std_offset`
+  changes along with the `zone_abbr` so a different period is needed during winter.
+  """
+  @type time_zone_period :: %{
+          optional(any) => any,
+          utc_offset: Calendar.utc_offset(),
+          std_offset: Calendar.std_offset(),
+          zone_abbr: Calendar.zone_abbr()
```

Comment:
I would suggest that we use a record or a raw tuple here. I know it's much less legible, but it's more performant and in things like this we probably want to squeeze every last bit of performance.

Reply:
I am not worried about eligibility but rather about extensibility. Using a tuple would be very unforgiving in case we need to add new information. A record is a bit better but adding custom fields would be tricky. Btw, if we have many maps, all with the same keys, do they share the keys structure when loading it in memory?

Reply:
It depends on how it is loaded. In general, though, if we're loading it from ETS, the table will always copy everything. ETS does not do literal sharing.

Reply:
I was wondering about modules. But keep in mind we can always store it in ets as a tuple and then convert it to a map when reading out. :)